x/erc20/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,172 @@
+	if sender.Equals(recipient) && !claimsParams.IsEVMChannel(packet.DestinationChannel) {
+		// Continue to the next IBC middleware by returning the original ACK.
+		return ack
+	}
+
+	// parse the transferred denom
+	coin := ibc.GetReceivedCoin(
+		packet.SourcePort, packet.SourceChannel,
+		packet.DestinationPort, packet.DestinationChannel,
+		data.Denom, data.Amount,
+	)
+
+	// check if the coin is a native staking token
+	bondDenom := k.stakingKeeper.BondDenom(ctx)
+	if coin.Denom == bondDenom {
+		// no-op, received coin is the staking denomination
+		return ack
+	}
+
+	// short-circuit: if the denom is not registered, conversion will fail
+	// so we can continue with the rest of the stack
+	if !k.IsDenomRegistered(ctx, coin.Denom) {
+		return ack
+	}
+
+	// Build MsgConvertCoin, from recipient to recipient since IBC transfer already occurred
+	msg := types.NewMsgConvertCoin(coin, common.BytesToAddress(recipient.Bytes()), recipient)
+
+	// NOTE: we don't use ValidateBasic the msg since we've already validated
+	// the ICS20 packet data
+
+	// Use MsgConvertCoin to convert the Cosmos Coin to an ERC20
+	if _, err = k.ConvertCoin(sdk.WrapSDKContext(ctx), msg); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	defer func() {
+		telemetry.IncrCounterWithLabels(
+			[]string{types.ModuleName, "ibc", "on_recv", "total"},
+			1,
+			[]metrics.Label{
+				telemetry.NewLabel("denom", coin.Denom),
+				telemetry.NewLabel("source_channel", packet.SourceChannel),
+				telemetry.NewLabel("source_port", packet.SourcePort),
+			},
+		)
+	}()
+
+	return ack
+}
+
+// OnAcknowledgementPacket responds to the the success or failure of a packet
+// acknowledgement written on the receiving chain. If the acknowledgement
+// was a success then nothing occurs. If the acknowledgement failed, then
+// the sender is refunded and then the IBC Coins are converted to ERC20.
+func (k Keeper) OnAcknowledgementPacket(
+	ctx sdk.Context, _ channeltypes.Packet,
+	data transfertypes.FungibleTokenPacketData,
+	ack channeltypes.Acknowledgement,
+) error {
+	switch ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Error:
+		// convert the token from Cosmos Coin to its ERC20 representation
+		return k.ConvertERC20AckPacket(ctx, data)
+	default:
+		// the acknowledgement succeeded on the receiving chain so nothing
+		// needs to be executed and no error needs to be returned
+		return nil
+	}
+}
+
+// OnTimeoutPacket converts the IBC coin to ERC20 after refunding the sender
+// since the original packet sent was never received and has been timed out.
+func (k Keeper) OnTimeoutPacket(ctx sdk.Context, _ channeltypes.Packet, data transfertypes.FungibleTokenPacketData) error {
+	return k.ConvertERC20AckPacket(ctx, data)
+}
+
+// ConvertERC20AckPacket converts the IBC coin to ERC20 after refunding the sender
+func (k Keeper) ConvertERC20AckPacket(ctx sdk.Context, data transfertypes.FungibleTokenPacketData) error {
```

Comment:
Consider: Function name contain `AckPacket` but It looks like function is not process/emit any acknowledgements. Maybe It can be renamed to reflect effect better? 
LMK your thoughts on this

Reply:
@4rgon4ut had the same thought here